docs/thinking_in_jax.rst

```diff
@@ -0,0 +1,484 @@
+You can think of :mod:`jax.lax` as a stricter, but often more powerful, API for working with
+multi-dimensional arrays.
+
+For example, :mod:`jax.numpy` will implicitly promote arguments to allow operations
+between mixed data types:
+
+.. ipython::
+
+   In [1]: import jax.numpy as jnp
+
+   @doctest
+   In [2]: jnp.add(1, 1.0)  # jax.numpy API implicitly promotes mixed types.
+   Out[2]: DeviceArray(2., dtype=float32)
+
+... and :mod:`jax.lax` will not:
+
+.. ipython::
+   
+   In [3]: from jax import lax
+
+   @verbatim
+   In [4]: lax.add(1, 1.0)  # jax.lax API requires explicit type promotion.
+   ---------------------------------------------------------------------------
+   TypeError                                 Traceback (most recent call last)
+   <ipython-input> in <module>
+   ----> 1 lax.add(1, 1.0)
+   ...
+   TypeError: add requires arguments to have the same dtypes, got int32, float32.
+
+If using :mod:`jax.lax` directly, you'll have to do type promotion explicitly in such cases:
+
+.. ipython::
+
+   @suppress
+   In [1]: from jax import lax
+
+   @doctest
+   In [1]: lax.add(jnp.float32(1), 1.0)
+   Out[1]: DeviceArray(2., dtype=float32)
+
+Along with this strictness, :mod:`jax.lax` also provides efficient APIs for some more
+general operations than are supported by NumPy.
+
+For example, consider a 1D convolution, which can be expressed in NumPy this way:
+
+.. ipython::
+
+   In [1]: x = jnp.array([1, 2, 1])
+
+   In [2]: y = jnp.ones(10)
+
+   In [3]: jnp.convolve(x, y)
+   Out[3]: DeviceArray([1., 3., 4., 4., 4., 4., 4., 4., 4., 4., 3., 1.], dtype=float32)
+
+Under the hood, this numpy operation is translated to a much more general convolution implemented by
+:func:`jax.lax.conv_general_dilated`:
+
+.. ipython::
+
+   In [1]: result = lax.conv_general_dilated(
+      ...:     x.reshape(1, 1, 3).astype(float),  # note: explicit promotion
+      ...:     y.reshape(1, 1, 10),
+      ...:     window_strides=(1,),
+      ...:     padding=[(len(y) - 1, len(y) - 1)])  # equivalent of padding='full' in numpy
+
+   @doctest
+   In [2]: result[0, 0]
+   Out[2]: DeviceArray([1., 3., 4., 4., 4., 4., 4., 4., 4., 4., 3., 1.], dtype=float32)
+
+This is a batched convolution operation designed to be efficient for the types of convolutions often
+used in deep neural nets. It requires much more boilerplate, but is far more flexible and scalable
+than the convolution provided by NumPy (See
+`JAX Sharp Bits: Convolutions <https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#%F0%9F%94%AA-Convolutions>`_
+for more detail on JAX convolutions).
+
+To JIT or not to JIT
+--------------------
+**Key Concepts:**
+
+- By default JAX executes operations one at a time, in sequence.
+- Using a just-in-time (JIT) compilation decorator, sequences of operations can be optimized
+  together and run at once.
+- Not all JAX code can be JIT compiled, as it requires array shapes to be static & known at
```

Comment:
Nit:

```suggestion
- Not all JAX code can be JIT compiled, as it requires array shapes to be static and known at
```

(`&` -> `and` like in `JIT mechanics: tracing and static variables`)